commit f1729c28a37e4f11ea5d9f468ab26adadb1aadab
Author: Don Mullis <dwm@meer.net>
Date:   Fri Dec 8 02:39:53 2006 -0800

    [PATCH] fault-injection: optimize and simplify should_fail()
    
    Trivial optimization and simplification of should_fail().
    
    Do cheaper disqualification tests first (performance gain not quantified).
    Simplify logic; eliminate goto.
    
    Signed-off-by: Don Mullis <dwm@meer.net>
    Cc: Akinobu Mita <akinobu.mita@gmail.com>
    Signed-off-by: Andrew Morton <akpm@osdl.org>
    Signed-off-by: Linus Torvalds <torvalds@osdl.org>

diff --git a/lib/fault-inject.c b/lib/fault-inject.c
index 5da665ac2d89..d143c0faf248 100644
--- a/lib/fault-inject.c
+++ b/lib/fault-inject.c
@@ -142,9 +142,6 @@ bool should_fail(struct fault_attr *attr, ssize_t size)
 	if (attr->task_filter && !fail_task(attr, current))
 		return false;
 
-	if (!fail_stacktrace(attr))
-		return false;
-
 	if (atomic_read(&attr->times) == 0)
 		return false;
 
@@ -159,12 +156,12 @@ bool should_fail(struct fault_attr *attr, ssize_t size)
 			return false;
 	}
 
-	if (attr->probability > random32() % 100)
-		goto fail;
+	if (attr->probability <= random32() % 100)
+		return false;
 
-	return false;
+	if (!fail_stacktrace(attr))
+		return false;
 
-fail:
 	fail_dump(attr);
 
 	if (atomic_read(&attr->times) != -1)